### ISSUE TITLE
[BUG]: Segmentation fault on Windows with abstract classss

### ISSUE THREAD

**Issue by TomWildenhain-Microsoft** [pybind/pybind11#3217]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

A have two abstract C++ class: Door and Knob.  Door returns a Knob with Door::GetKnob.  The knob can be pulled with Knob::Pull().  Door successfully calls Door::GetKnob and gets a non-null knob, but the code crashes when Knob::Pull is called.

### Reproducible example code

C++
```C++
#include <iostream>
#include <pybind11/pybind11.h>

namespace py = pybind11;

int simple_add(int i, int j) {
    return i + j;
}

class Knob {
 public:
  virtual void Pull() = 0;
};

class Door {
 public:
  virtual std::string_view Name() = 0;
  virtual Knob& GetKnob() = 0;
};

void open_door(Door& door) {
  std::string_view name = door.Name();
  std::cout << name << std::endl;
  Knob& knob = door.GetKnob();
  std::cout << "Got the knob" << std::endl;
  knob.Pull();
  std::cout << "Pulled the knob" << std::endl;
}

class PyKnob : public Knob {
 public:
  using Knob::Knob;
  void Pull() {
    PYBIND11_OVERRIDE_PURE(void, Knob, Pull);
  }
};

class PyDoor : public Door {
 public:
  using Door::Door;
  std::string_view Name() {
    PYBIND11_OVERRIDE_PURE(std::string_view, Door, Name);
  }
  Knob& GetKnob() {
    PYBIND11_OVERRIDE_PURE(Knob&, Door, GetKnob);
  }
};

PYBIND11_MODULE(pybind_test, m) {
  m.def("simple_add", &simple_add);
  m.def("open_door", &open_door);
  py::class_<Knob, PyKnob>(m, "Knob")
    .def(py::init<>())
    .def("Pull", &Knob::Pull);
  py::class_<Door, PyDoor>(m, "Door")
    .def(py::init<>())
    .def("Name", &Door::Name)
    .def("GetKnob", &Door::GetKnob);
}
```

Python
```python
import pybind_test

print(pybind_test.simple_add(3, 4))

class MyKnob(pybind_test.Knob):
    def __init__(self):
        pybind_test.Knob.__init__(self)

    def Pull(self):
        print("Hey!")

class MyDoor(pybind_test.Door):
    def __init__(self, name):
        pybind_test.Door.__init__(self)
        self.name = name

    def Name(self):
        return self.name
    
    def GetKnob(self):
        return MyKnob()

print("Opening")
pybind_test.open_door(MyDoor("I'm a door."))
print("Done")
```

Output
```
7
Opening
I'm a door.
Got the knob
```
Program crashes at this point. (while calling `knob.Pull();`)

**Comment by laramiel** [comment]
Try allocating the knob in the `MyDoor` constructor:

```py
class MyDoor(pybind_test.Door):
    def __init__(self, name):
        pybind_test.Door.__init__(self)
        self.name = name
        self.knob = MyKnob()

    def Name(self):
        return self.name
    
    def GetKnob(self):
        return self.knob
```

Then at least you're not returning a temporary.

**Comment by TomWildenhain-Microsoft** [comment]
@laramiel yes that works but isn't the behavior I want. Pretend it's called "Knob factory" instead ;-).  Also I tried shared_pointers and return value policies and couldn't get it to stop deallocating on return.  The failure of shared_pointer really surprised me.

**Comment by henryiii** [comment]
I've already said - you can't return a reference to a temporary object with a virtual class. It doesn't work. The compiler slices off part of your object. It's UB.

Using shared pointers should be just fine. Do you have the shared pointer version to look at?

**Comment by TomWildenhain-Microsoft** [comment]
> Using shared pointers should be just fine. Do you have the shared pointer version to look at?

Thanks for your help.  Replace all  `Knob&` with  `std::shared_ptr<Knob>` to get the shared pointer version.

**Comment by TomWildenhain-Microsoft** [comment]
The cause of the issue is described here:
https://github.com/pybind/pybind11/issues/673#issuecomment-280828162

> You were on the right track with shared_ptr. The reason you get the error about calling a virtual call() is that the Python object doesn't exist anymore at the time you're making the call. You still have the C++ instance, in the shared_ptr, but the Python instance (which was returned by A.create()) isn't around anymore because nothing references it once the call is over: pybind11 itself only holds on to it long enough to convert the value to a C++ type (std::shared_ptr<Animal> in this case), which means that in your testing function, your t will be a shared pointer to a PyAnimal, but when pybind11 tries to look up the python instance associated with t to find the overridden call function there is no Python instance anymore.

It would be great if pybind could keep the python instance around when there is a shared pointer to it.
